File: python/doubly-llist.py
""" Doubly Linked List
NODE: [prev|data|next]
Each node stores the data and the address of the next node

HEAD --> [prev|data|next] <-> [prev|data|next] <-> [prev|data|next] -> NULL

Basic Operations:
    prepend(data)       - Adds an element at the beginning of the list.
    append(data)        - Appends an element at the end of the list
    pop_first()         - Deletes an element at the beginning of the list.
    pop_last            - Deletes an element from the end of the list.
    insert_after        - Adds an element after an item of the list.
    delete(data)        - Deletes an element from the list using the key.
    display_forward()   - Displays the complete list in a forward manner.
    display_backward()  - Displays the complete list in a backward manner.
"""


class Node(object):
    def __init__(self, data, prev=None, next=None):
        self.data = data
        self.prev = prev
        self.next = next

    def __repr__(self) -> str:
        # return "Node(data={self.data}, prev={self..data}, next={self.next.data})"
        return print(f"Node({self.data})")


class DLinkedList(object):
    def __init__(self):
        self.head = None
        self.tail = None
        self.count = 0

    # def __repr__(self):
    #     curr = self.head
    #     nodes = "\n"
    #     while curr is not None:
    #         nodes += repr(curr) + " --> "
    #         curr = curr.next
    #     nodes += "None\n"
    #     return nodes

    def __repr__(self) -> str:
        curr = self.head
        nodes = []
        while curr is not None:
            nodes.append("[" + str(curr.data) + "]")
            curr = curr.next
        nodes.append("None")
        return " --> ".join(nodes)

    def __str__(self):
        """Prints the node data from the list"""
        nodes = []
        llist = "["
        curr = self.head
        while curr and curr.next is not None:
            nodes.append(curr.data)
            curr = curr.next
        # return ", ".join(map(str, nodes))
        llist += "] --> [".join(map(str, nodes)) + "]"
        return llist

    def is_empty(self):
        if self.head is None:
            return True
        else:
            return False
        # return (self.head is None) ? True: False
        # return True if self.head is None else False

    def fetch_tail(self):
        """Returns the last node in the list"""
        curr = self.head
        while curr and curr.next is not None:
            curr = curr.next
        return curr

    def append(self, data):
        """Appends new node to end of list"""
        node = Node(data)
        if self.is_empty():
            self.head = node
        else:
            tail = self.fetch_tail()  # current last
            tail.next = node
            self.tail = node
            self.tail.prev = tail
            self.tail.next = None
        self.count += 1

    def prepend(self, data):
        """Adds node to beginning of list"""
        node = Node(data)
        if self.is_empty():
            self.head = node
        else:
            node.prev = None
            node.next = self.head
            self.head = node
        self.count += 1

    def pop_first(self):
        """Removes the first node from the list.
           Returns the data of the last node removed."""
        if not self.is_empty():
            temp = self.head
            self.head = self.head.next
            self.count -= 1
            print(f"\nNode({temp.data}) was removed from the list")
            return temp.data  # returns the last data removed
        return None

    def pop_last(self):
        """Removes the last node from the list.
           Returns the data of the last node removed."""
        if not self.is_empty():
            prev = None
            curr = self.head
            while curr and curr.next is not None:
                prev = curr
                curr = curr.next
            prev.next = None
            self.tail = prev
            self.count -= 1
            print(f"\nNode({curr.data}) was removed from the list")
            return curr.data  # returns the last data removed
        return None

    def insert_after(self, item, data):
        """Adds an new node after an node:item of the list."""
        prev = None
        curr = self.head
        while curr and curr.next is not None:
            prev = curr
            curr = curr.next
            if (prev.next.data == item.data and curr.data == item.data):
                # node = Node(data)
                # node.prev = prev
                # node.next = curr
                node = Node(data, prev=prev, next=curr)
                prev.next = node
                curr.prev = node
                self.count += 1
                print(f"\nNode({data}) was inserted after Node({item.data})")
                return True
        print(f"\nNode({item}) not found in list")
        return False

    def remove(self, item):
        """Removes a node:item from the list using the key:data.
           Returns True if the node was removed"""
        prev = None
        curr = self.head
        while curr and curr.next is not None:
            prev = curr
            curr = curr.next
            if (curr.data == item.data):
                prev.next = curr.next
                curr.next.prev = prev
                self.count -= 1
                print(f"\nNode({item.data}) was removed")
                return True
        print(f"\nNode({item.data}) could not be removed")
        return False

    def display(self, reverse=False):
        nodes = ["HEAD"]
        curr = self.head
        while curr is not None:
            nodes.append(curr.data)
            curr = curr.next
        nodes.append("NONE")
        if reverse:
            nodes.reverse()
        return " --> ".join(map(str, nodes))

    # def display_forward(self):
    #     """Displays the complete list in a foward manner"""
    #     nodes = ["HEAD"]
    #     curr = self.head
    #     while curr is not None:
    #         nodes.append(curr.data)
    #         curr = curr.next
    #     nodes.append("NONE")
    #     return " --> ".join(map(str, nodes))

    # def display_backward(self):
    #     """Displays the complete list in a backward manner"""
    #     nodes = ["HEAD"]
    #     curr = self.head
    #     while curr is not None:
    #         nodes.append(curr.data)
    #         curr = curr.next
    #     nodes.append("NONE")
    #     nodes.reverse()
    #     return " --> ".join(map(str, nodes))


if __name__ == "__main__":
    llist = DLinkedList()
    llist.append(1)
    llist.append(2)
    llist.append(3)
    llist.prepend(0)
    llist.append(4)
    llist.prepend(-1)
    # print(llist.display_forward())
    # print(llist.display_backward())
    print(llist.display())
    print(llist.display(reverse=True))
    llist.remove(Node(-100))
    print(llist.display())

    llist.remove(Node(0))
    print(llist.display())

    llist.insert_after(Node(2), 100)
    print(llist.display())

    llist.pop_first()
    print(llist.display())

    llist.pop_last()
    print(llist.display())

    print(repr(llist))
    print(str(llist))


# map(function, iterable) : maps a function to each item in the iterable
# delimter.join(iterable) : joins every item in the iterable into a single delimited string

File: c/doubly-llist.c
/** Doubly Linked List */
#include <stdio.h> // required for NULL
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <stdbool.h> // required for bool type

struct node
{
    int key;
    int data;
    struct node *next;
    struct node *prev;
};

struct node *head = NULL;
struct node *tail = NULL;
struct node *curr = NULL;

bool isEmpty()
{
    return head == NULL;
}

int length()
{
    int length = 0;
    struct node *curr;

    for (curr = head; curr != NULL; curr = curr->next)
    {
        length++;
    }
    return length;
}

void displayFoward()
{
    printf("\n[ ");

    struct node *ptr = head; // start at the head
    while (ptr != NULL)
    {
        printf("(%d, %d)", ptr->key, ptr->data);
        ptr = ptr->next;
    }
    printf("");
}

void displayBackward()
{
    printf("\n[ ");

    struct node *ptr = tail; // start at the tail
    while (ptr != NULL)
    {
        printf("(%d, %d) ", ptr->key, ptr->data);

        ptr = ptr->prev;
    }
}

// Insert node at the beginning of the list
void insertFirst(int key, int data)
{
    struct node *node = (struct node *)malloc(sizeof(struct node));
    node->key = key;
    node->data = data;
    // (*node).key = key;
    // (*node).data = data;

    if (isEmpty())
    {
        tail = node;
    }
    else
    {
        head->prev = node;
    }

    // Update pointers and set new head of the list
    node->next = head;
    head = node;
}

// Insert node at the end of the list
void insertLast(int key, int data)
{
    struct node *node = (struct node *)malloc(sizeof(struct node));
    node->key = key;
    node->data = data;

    if (isEmpty())
    {
        // Set the tail node
        tail = node;
    }
    else
    {
        // Update pointers and set new tail node
        tail->next = node;
        node->prev = tail;
    }

    tail = node;
}

struct node *deleteFirst()
{
    // Save reference to former head
    struct node *temp = head;

    // If only one node
    if (head->next == NULL)
    {
        tail = NULL;
    }
    else
    {
        head->next->prev = NULL;
    }

    head = head->next;
    return temp; // return the deleted node
}

struct node *deleteLast()
{
    // save reference to last node
    struct node *temp = tail;

    // if only one link
    if (head->next == NULL)
    {
        head = NULL;
    }
    else
    {
        tail->prev->next = NULL;
    }

    tail = tail->prev;

    return temp;
}

// Delete a node with given key
struct node *delete(int key)
{

    // start from the first link
    struct node *current = head;
    struct node *previous = NULL;

    // if list is empty
    if (head == NULL)
    {
        return NULL;
    }

    // navigate through list
    while (current->key != key)
    {
        // if it is last node

        if (current->next == NULL)
        {
            return NULL;
        }
        else
        {
            // store reference to current link
            previous = current;

            // move to next link
            current = current->next;
        }
    }

    // found a match, update the link
    if (current == head)
    {
        // change first to point to next link
        head = head->next;
    }
    else
    {
        // bypass the current link
        current->prev->next = current->next;
    }

    if (current == tail)
    {
        // change last to point to prev link
        tail = current->prev;
    }
    else
    {
        current->next->prev = current->prev;
    }

    return current;
}

bool insertAfter(int key, int newKey, int data)
{
    // start from the first link
    struct node *current = head;

    // if list is empty
    if (head == NULL)
    {
        return false;
    }

    // navigate through list
    while (current->key != key)
    {

        // if it is last node
        if (current->next == NULL)
        {
            return false;
        }
        else
        {
            // move to next link
            current = current->next;
        }
    }

    // create a link
    struct node *newLink = (struct node *)malloc(sizeof(struct node));
    newLink->key = newKey;
    newLink->data = data;

    if (current == tail)
    {
        newLink->next = NULL;
        tail = newLink;
    }
    else
    {
        newLink->next = current->next;
        current->next->prev = newLink;
    }

    newLink->prev = current;
    current->next = newLink;
    return true;
}

void main()
{
    insertFirst(1, 10);
    insertFirst(2, 20);
    insertFirst(3, 30);
    insertFirst(4, 1);
    insertFirst(5, 40);
    insertFirst(6, 56);

    printf("\nList (First to Last): ");
    displayForward();

    printf("\n");
    printf("\nList (Last to first): ");
    displayBackward();

    printf("\nList , after deleting first record: ");
    deleteFirst();
    displayForward();

    printf("\nList , after deleting last record: ");
    deleteLast();
    displayForward();

    printf("\nList , insert after key(4) : ");
    insertAfter(4, 7, 13);
    displayForward();

    printf("\nList  , after delete key(4) : ");
    delete (4);
    displayForward();
}

// https://www.tutorialspoint.com/data_structures_algorithms/doubly_linked_list_program_in_c.htm
